Multiple master apps for linked apps

orangejenny opened the pull request:
@mkangia @esoergel 

Opening for review/reference before I go offline. The first handful of commits is from https://github.com/dimagi/commcare-hq/pull/24356 with the feedback from that PR rebased in. The first new commit is daec325

This PR adds the ability to pick a master app when pulling a linked app:
![Screen Shot 2019-06-18 at 8 08 43 PM](https://user-images.githubusercontent.com/1486591/59727880-a0191180-9205-11e9-94c1-a846ddb4a0b1.png)

The releases page shows where each version came from:
![Screen Shot 2019-06-18 at 8 10 11 PM](https://user-images.githubusercontent.com/1486591/59727896-a9a27980-9205-11e9-8596-e4d0ff7f9cb3.png)

This does:
- allow updating a linked app (like changing one of the overridable app settings), then saving, then making a new build
- handle form unique ids properly (they're consistent between all linked app versions that pull from the same master)
- make linked apps support reverts like normal apps (creates a new build that's a copy of the build being reverted to)

It does not
- have any new tests, and I'm sure it breaks a bunch of existing tests 🙈 
- correctly version forms or multimedia

orangejenny committed:
Added pulled_from_master_version to LinkedApplication

orangejenny committed:
Added pulled_from_master_version to saved_app view

orangejenny committed:
Added UI indicator for pulled_from_master_version to releases table

orangejenny committed:
Added pulled_from_master_app_version to ES

orangejenny committed:
Added test for incremental versioning

orangejenny committed:
Added default logic for pulled_from_master_version in couch view

orangejenny committed:
Added pulled_from_master_app_id attribute and reflected it in UI.

orangejenny committed:
Updated application saving to increment version even for linked apps.

This allows linked apps to create new builds after changing
independent/overridable settings like CommCare flavor.

orangejenny committed:
Added corehq.apps.linked_domain.applications.get_master_app_briefs

orangejenny committed:
Replaced master app ids on releases page with master app names and links

orangejenny committed:
Removed LinkedApplication.get_master_version

orangejenny committed:
Updated masters dropdown to display most recently released version

orangejenny committed:
Fixed form unique ids changing between versions

Form unique ids were changing between linked app builds pulled from
the same master if there's a build from a different master between them.
The pull logic looks at the XMLNSes in the previous version's forms
and assigns a new unique id if it doesn't find the XMLNS, so fixing this
is just a matter of looking at the previous version from the same master
rather than the linked app's immediate predecessor.

orangejenny committed:
Fixed lint errors.

Clearly this hasn't been tested with remote linked apps yet.
The naming of the various versions by id functions still needs work.

orangejenny committed:
Updated form unique ID logic to update all IDs on the first pull of a new master.

orangejenny commented on corehq/apps/app_manager/views/utils.py:
```diff
@@ -188,27 +187,20 @@ def overwrite_app(app, master_build, report_map=None):
     app_source.pop('external_blobs')
     app_source['_attachments'] = _attachments
 
+    form_ids_by_xmlns = _get_form_ids_by_xmlns(app.get_previous_version(master_app_id=master_app.copy_of))
```
Added comment in 69ea917ee91b3e0048fb644ebefa389c51db6687 let me know if that clarifies - and if not, let's offline, I'm not 100% following your specific questions.

orangejenny replied:
Also added 9ab52a3eaebf9bd186b2b9c59cf11c3853287976 to update *all* form unique ids the first time that a new master is pulled. This mimics prod behavior, where when a linked app is created, `_update_form_ids` gets called with `form_ids_by_xmlns` an empty dict.

esoergel replied:
Yeah, let's offline, I'm still not convinced, though I may be misunderstanding - think I need to talk through some examples.

orangejenny replied:
offlined, resolving

orangejenny committed:
Added logic to handle form unique ids when the same form has been added to multiple masters

orangejenny committed:
Removed to-be-deprecated master property from test_linked_apps.py

orangejenny committed:
Added test for new upstream_* attributes

orangejenny committed:
Fixed lint errors

orangejenny committed:
Added multi-master coverage to test_get_latest_master_release

orangejenny committed:
Cleaned up test_linked_apps.py

Renamed some things to be more concise and remove from redundant saves
and redundant cleanup.

emord commented on corehq/apps/app_manager/dbaccessors.py:
```diff
@@ -300,21 +316,35 @@ def get_built_app_ids(domain):
+
+def get_build_ids(domain, app_id):
+    """
+    Returns all the built apps for an application id, in descending order.
```
Is this descending by app_id, or time built?

orangejenny replied:
Clarified in a31ee03e63d although given that app ids aren't numeric I don't think it's likely a reader will think that's the sort order.

emord commented on corehq/apps/app_manager/models.py:
```diff
@@ -5599,21 +5602,50 @@ def domain_link(self):
-            return get_latest_master_app_release(self.domain_link, self.master)
+            return get_latest_master_app_release(self.domain_link, master_app_id)
         else:
             raise ActionNotPermitted
```
I realize that you didn't write this, but the if/else doesn't match the rest of the code in this class. Could be

```
if self.domain_link:
   # do thing
raise ActionNotPermitted
```

orangejenny replied:
46e31c48c1c

emord commented on corehq/apps/app_manager/tasks.py:
```diff
@@ -67,9 +67,9 @@ def prune_auto_generated_builds(domain, app_id):
 
 @task(serializer='pickle', queue='background_queue', ignore_result=True)
-def update_linked_app_and_notify_task(domain, app_id, user_id, email):
+def update_linked_app_and_notify_task(domain, app_id, master_app_id, user_id, email):
```
This would break any linked app tasks that are in the queue during deploy since they wouldn't provide the `master_app_id`. There's probably few enough of these calls for that to be fine, but wanted to flag in case its unintentional

orangejenny replied:
Ugh, you're right. I'm not planning to merge this PR as-is, but I added a note to the rollout plan to handle the param change - https://docs.google.com/a/dimagi.com/document/d/1dXiVcm871UQIqi5-tFD7rTcLemBxxikpaCiFr6gmkLI/edit?disco=AAAADO5txdY

emord commented on corehq/apps/app_manager/views/apps.py:
```diff
@@ -280,19 +285,29 @@ def _get_setting(setting_type, setting_id):
         context['upstream_url'] = _get_upstream_url(app, request.couch_user)
         try:
-            context['master_version'] = app.get_master_version()
+            master_versions_by_id = get_latest_released_app_versions_by_app_id(app.domain)
```
what part(s) of this try/except block could raise `RemoteRequestError`? could this be changed to a try/catch/else?

orangejenny replied:
c0c3d253fed

emord commented on corehq/apps/linked_domain/tests/test_linked_apps.py:
```diff
@@ -39,10 +39,15 @@ def setUpClass(cls):
+        cls.master2 = Application.new_app(cls.domain, "Second Master Application")
+        cls.linked_domain = 'domain-2'
+        cls.master2.linked_whitelist = [cls.linked_domain]
+        cls.master2.save()
```
Does this test change imply that we're expecting all linked apps to have access to multi-master? If not, would it be useful to have some explicit single-master tests around?

orangejenny replied:
Expecting all linked apps to have access to multi-master.

emord commented on corehq/apps/app_manager/models.py:
```diff
@@ -5577,8 +5579,9 @@ class LinkedApplication(Application):
-    # This is the id of the master application
-    master = StringProperty()
+    master = StringProperty()   # Legacy, should be removed once all linked apps support multiple masters
+    upstream_app_id = StringProperty()      # ID of the app that was most recently pulled
```
What's the difference between master and upstream? When should we say master vs upstream or will they be used interchangeably?

esoergel replied:
Jenny and I have been discussing this.  Here's my take on the terminology:

In the current world each linked app has a single "master" app.  In the future world, each linked app will be able to pull from multiple "upstream" apps, all of which descend from a single "progenitor" app.

What do you think?

emord replied:
Interesting. Are there going to be any constraints on enforcing the same "progenitor" app? Does that mean that master as a linked app term will never be used after this PR migrates current linked apps?

I'm mostly asking for my own understanding as we've overloaded "master" when talking about this in the past. As long as we're clear on the terminology being used, I'll probably be happy with the outcome

esoergel replied:
> Does that mean that master as a linked app term will never be used after this PR migrates current linked apps?

Yeah, that's the idea.  The "master" terminology is for sure ambiguous (`master_id` is an unrelated concept on the same model), so hopefully this'll clear it up.

orangejenny replied:
I'd like to wholly get rid of "master" as it relates to linked apps, but that's a rather large change and would also affect linked domains. My major complaint is the one Ethan mentions: in current production, `VersionedDoc` has a `master_id` property, `LinkedApplication` has a `master` property, and while both of these are strings that contain an app id, they have nothing to do with each other. Come to think of it, it might be easier to rename `master_id` (not sure to what...`canonical_id`?) and let linked apps/projects keep "master," rather than introducing "upstream" - @esoergel what do you think?

orangejenny committed:
Extracted _pull_linked_app function

orangejenny committed:
Added test of form attributes when pulling from several master apps

orangejenny committed:
Added test of adding same form to multiple master apps

orangejenny committed:
Removed redundant else

orangejenny committed:
Limited scope of try/catch block

orangejenny committed:
Added test for multimedia versioning

orangejenny commented:
Added tests and responded to feedback. Planning to do any future changes in spinoff PRs.

orangejenny committed:
Limited master apps dropdown to apps with multiple masters

orangejenny commented:
Rebased and force pushed. Dropped the commits that were cherry-picked into https://github.com/dimagi/commcare-hq/pull/25003.

orangejenny committed:
Added family_id

Added progenitor_app_id when copying app via UI

Limited which apps can be pulled to those with the same progenitor

Added tests for progenitor_app_id

Added logic to check previous progenitor build if no previous build exists

Make arg required and rename to disambiguate

orangejenny committed:
Added tests for Application.get_previous_version and LinkedApplication.get_previous_version

orangejenny committed:
Added support for pulling from multiple masters

orangejenny committed:
Renamed get_built_app_ids_for_app_id and get_built_app_ids_for_app_id_descending

orangejenny committed:
Added test for get_latest_released_app_versions_by_app_id

orangejenny commented:
Closing in favor of https://github.com/dimagi/commcare-hq/pull/25107
